File: nodejs/agents/okx.ts
import * as crypto from "crypto";
import { ExecutionAgent, OrderDetails } from "./base";

function signHmacSha256(secret: string, data: string): string {
    return crypto.createHmac("sha256", secret).update(data).digest("hex");
}

export class OkxAgent extends ExecutionAgent {
    private apiKey: string;
    private apiSecret: string;
    private passphrase: string;
    private apiUrl: string;

    constructor(
        name: string,
        symbolMap: Record<string, string>,
        apiKey: string,
        apiSecret: string,
        passphrase: string,
        apiUrl: string
    ) {
        super(name, symbolMap);
        this.apiKey = apiKey;
        this.apiSecret = apiSecret;
        this.passphrase = passphrase;
        this.apiUrl = apiUrl;
    }

    public async execute(event: any, details: OrderDetails): Promise<void> {
        if (details.quantity <= 0) {
            return;
        }

        const resolvedSymbol = this.symbolMap[details.symbol];
        const timestamp = new Date().toISOString();
        const path = "/api/v5/trade/order";
        const url = `${this.apiUrl}${path}`;
        
        const bodyData: Record<string, any> = {
            instId: resolvedSymbol,
            tdMode: "cash",
            side: details.side.toLowerCase(),
            ordType: details.orderType.toLowerCase(),
            sz: String(details.quantity)
        };
        
        if (details.orderType.toUpperCase() === "LIMIT") {
            if (details.price !== undefined && details.price > 0) {
                bodyData.px = String(details.price);
            } else {
                console.error(`Error [OKX Agent '${this.name}']: Price required for LIMIT orders`);
                return;
            }
        }
        
        const bodyStr = JSON.stringify(bodyData);
        const signPayload = `${timestamp}POST${path}${bodyStr}`;
        const signature = signHmacSha256(this.apiSecret, signPayload);
        
        console.log(`Placing OKX order via Agent '${this.name}': ${details.side} ${resolvedSymbol} (qty: ${details.quantity})`);
        
        try {
            const res = await fetch(url, {
                method: "POST",
                headers: {
                    "OK-ACCESS-KEY": this.apiKey,
                    "OK-ACCESS-SIGN": signature,
                    "OK-ACCESS-TIMESTAMP": timestamp,
                    "OK-ACCESS-PASSPHRASE": this.passphrase,
                    "Content-Type": "application/json"
                },
                body: bodyStr
            });
            const body = await res.text();
            if (res.ok) {
                console.log(`OKX Order via Agent '${this.name}' Succeeded:`, body);
            } else {
                console.error(`OKX Order via Agent '${this.name}' Failed (HTTP ${res.status}):`, body);
            }
        } catch (err) {
            console.error(`OKX Connection Error via Agent '${this.name}':`, err);
        }
    }
}

File: rust/src/agents/okx.rs
use crate::agents::{ExecutionAgent, MappedOrder};
use std::collections::HashMap;
use tickengine_sdk::ClientEvent;
use tracing::{error, info};

pub struct OkxAgent {
    name: String,
    client: reqwest::Client,
    api_key: String,
    api_secret: String,
    passphrase: String,
    api_url: String,
    symbol_map: HashMap<String, String>,
}

impl OkxAgent {
    pub fn new(
        name: &str,
        client: reqwest::Client,
        api_key: &str,
        api_secret: &str,
        passphrase: &str,
        api_url: &str,
        symbol_map: HashMap<String, String>,
    ) -> Self {
        Self {
            name: name.to_string(),
            client,
            api_key: api_key.to_string(),
            api_secret: api_secret.to_string(),
            passphrase: passphrase.to_string(),
            api_url: api_url.to_string(),
            symbol_map,
        }
    }
}

fn sign_hmac_sha256(secret: &str, data: &str) -> String {
    use hmac::{Hmac, Mac};
    use sha2::Sha256;
    type HmacSha256 = Hmac<Sha256>;

    let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).expect("HMAC keys can be any size");
    mac.update(data.as_bytes());
    hex::encode(mac.finalize().into_bytes())
}

#[async_trait::async_trait]
impl ExecutionAgent for OkxAgent {
    fn name(&self) -> &str {
        &self.name
    }

    fn can_handle(&self, symbol: &str) -> bool {
        self.symbol_map.contains_key(symbol)
    }

    async fn execute(&mut self, _event: &ClientEvent, order: &MappedOrder) -> anyhow::Result<()> {
        let resolved_symbol = self
            .symbol_map
            .get(&order.symbol)
            .ok_or_else(|| anyhow::anyhow!("Symbol not mapped for OKX agent"))?;

        if order.quantity <= 0.0 {
            return Ok(());
        }

        let timestamp = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let path = "/api/v5/trade/order";
        let url = format!("{}{}", self.api_url, path);

        let mut body_map = serde_json::Map::new();
        body_map.insert(
            "instId".to_string(),
            serde_json::Value::String(resolved_symbol.clone()),
        );
        body_map.insert(
            "tdMode".to_string(),
            serde_json::Value::String("cash".to_string()),
        );
        body_map.insert(
            "side".to_string(),
            serde_json::Value::String(order.side.to_lowercase()),
        );
        body_map.insert(
            "ordType".to_string(),
            serde_json::Value::String(order.order_type.to_lowercase()),
        );
        body_map.insert(
            "sz".to_string(),
            serde_json::Value::String(order.quantity.to_string()),
        );

        if order.order_type == "LIMIT" {
            if let Some(p) = order.price {
                body_map.insert("px".to_string(), serde_json::Value::String(p.to_string()));
            } else {
                return Err(anyhow::anyhow!("Price required for OKX LIMIT orders"));
            }
        }

        let body_str = serde_json::to_string(&serde_json::Value::Object(body_map))?;
        let sign_payload = format!("{}{}{}{}", timestamp, "POST", path, body_str);
        let signature = sign_hmac_sha256(&self.api_secret, &sign_payload);

        info!(
            "Placing OKX order via Agent '{}': {} {} (qty: {})",
            self.name, order.side, resolved_symbol, order.quantity
        );

        let res = self
            .client
            .post(&url)
            .header("OK-ACCESS-KEY", &self.api_key)
            .header("OK-ACCESS-SIGN", signature)
            .header("OK-ACCESS-TIMESTAMP", timestamp)
            .header("OK-ACCESS-PASSPHRASE", &self.passphrase)
            .header("Content-Type", "application/json")
            .body(body_str)
            .send()
            .await?;

        let status = res.status();
        let body = res.text().await?;

        if status.is_success() {
            info!("OKX Order via Agent '{}' Succeeded: {}", self.name, body);
        } else {
            error!(
                "OKX Order via Agent '{}' Failed (HTTP {}): {}",
                self.name, status, body
            );
        }

        Ok(())
    }
}
